### PR TITLE
Feature Request: Provide a way to use BlueCoat Sitereview API

### PR DESCRIPTION
**Pull request by cgi1**
+ baseLib.py (a base class to use a custom configuration and logging)
+ BlueCoatAPI.py (provides the ability to submit URLs for (re-)categorization using an API access to BlueCoats SiteReview service.)
+ BlueCoatAPI.conf (configuration for BlueCoatAPI.py)

### CODE REVIEW

**Comment by cgi1** [comment]
Solves #969

**Comment by wagner-certat** [comment]
Should this be a bot? I only see a configuration file and two library files.

**Comment by cgi1** [comment]
Forget the expert (see code below). I can create an additional Feature Request for it?

```
# -*- coding: utf-8 -*-

import sys

from intelmq.lib import BlueCoatAPI
from intelmq.lib.bot import Bot

__author__ = 'Christoph Giese <cgi1> <C.Giese@telekom.de>'

class BlueCoatMitigationBot(Bot):

    def init(self):
        self.logger.info("Init of BlueCoat Mitigation Bot started...")
        self.bluecoat_api = BlueCoatAPI.BlueCoatAPI()
        self.logger.info("Init of BlueCoat Mitigation Bot finished!")

    def __get_mititgation_value(self, event):
        """ return the URL/IP/Domain which needs to be mitigated from one of the different event fields

        :param event:
        :return: string value
        """
        return event.get('source.ip', None) or \
               event.get('destination.ip', None) or \
               event.get('source.url', None) or \
               event.get('destination.url', None) or \
               event.get('source.fqdn', None) or \
               event.get('destination.fqdn', None)

    def process(self):
        """ Receives an event which contains a mitigation request (i.e. block at bluecoat proxy)

        :return: Nothing
        """

        event = self.receive_message()
        # Check if mitigation command is set
        if 'control.command' in event:
            self.logger.info("Received command: %s" % (event['control.command']))
            if event['control.command'] == 'mitigate':
                mitigation_value = self.__get_mititgation_value(event)
                response_code = self.bluecoat_api.submit_url(
                    url=mitigation_value,
                    submit_comment=self.parameters.submit_comment,
                    cat1=self.parameters.bc_category1,
                    cat2=self.parameters.bc_category2,
                    email=self.parameters.submitter_mail,
                    customTrackingID=123,
                    confidence=80)
                self.logger.info("Response code from BlueCoat: %s." % response_code)
                if response_code:
                    event.add('control.response_code', response_code)
                else:
                    event.add('control.response_code', 666)
                event.add('control.raw_response', response_code)
                self.send_message(event)
        else:
            self.logger.debug("Received event without control.command.")
        self.acknowledge_message()

if __name__ == "__main__":
    bot = BlueCoatMitigationBot(sys.argv[1])
    bot.start()
```

**Comment by wagner-certat** [comment]
Please just add it to this PR

File: intelmq/bots/experts/bluecoat/expert_bluecoat.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+                    event.add('control.response_code', response_code)
+                else:
+                    event.add('control.response_code', 666)
+                event.add('control.raw_response', response_code)
+                self.send_message(event)
+        else:
+            self.logger.debug("Received event without control.command.")
+        self.acknowledge_message()
+
+if __name__ == "__main__":
```
Comment (wagner-certat) [inline intelmq/bots/experts/bluecoat/expert_bluecoat.py]: Replace this block with `BOT = BlueCoatMitigationBot`

File: intelmq/bots/experts/bluecoat/expert_bluecoat.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
+# -*- coding: utf-8 -*-
+
+import sys
+
+from intelmq.lib import BlueCoatAPI
+from intelmq.lib.bot import Bot
+
+__author__ = 'Christoph Giese <cgi1> <C.Giese@telekom.de>'
+
+class BlueCoatMitigationBot(Bot):
```
Comment (wagner-certat) [inline intelmq/bots/experts/bluecoat/expert_bluecoat.py]: rename to `BlueCoatMitigationExpertBot`

File: intelmq/bots/experts/bluecoat/expert_bluecoat.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+    def init(self):
+        self.logger.info("Init of BlueCoat Mitigation Bot started...")
```
Comment (wagner-certat) [inline intelmq/bots/experts/bluecoat/expert_bluecoat.py]: remove this line

File: intelmq/bots/experts/bluecoat/expert_bluecoat.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        self.bluecoat_api = BlueCoatAPI.BlueCoatAPI()
+        self.logger.info("Init of BlueCoat Mitigation Bot finished!")
```
Comment (wagner-certat) [inline intelmq/bots/experts/bluecoat/expert_bluecoat.py]: remove this line

**Comment by wagner-certat** [comment]
> @wagner-certat So you do not want to have logging into the libaries? That does not fit for this case imo.

In case you refer to https://github.com/certtools/intelmq/pull/969#discussion_r115702118 and https://github.com/certtools/intelmq/pull/969#discussion_r115702154: The Bot class itself is logging about the initialization

File: intelmq/lib/BlueCoatAPI.py [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+        headers = {
+            'content-type': 'application/x-www-form-urlencoded'
+        }
+
+        try:
+            parsed_response = None
+            response = requests.post(self.report_threat_url, data=payload, headers=headers, verify=False)
+            if response.status_code == 200:
+                parsed_response = xmltodict.parse(response._content)
+                self.logger.debug("Successfully parsed response from BlueCoat-Submit for URL [" + str(url) + "]:" + str(
```
Comment (cgi1) [inline intelmq/lib/BlueCoatAPI.py]: @wagner-certat: This logging is only possible with the baseLib.

File: intelmq/lib/BlueCoatAPI.py [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+                    parsed_response))
+            else:
+                self.logger.error("Error in response from BlueCoat API")
```
Comment (cgi1) [inline intelmq/lib/BlueCoatAPI.py]: @wagner-certat: This logging is only possible with the baseLib.

File: intelmq/lib/BlueCoatAPI.py [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+        if parsed_response and len(parsed_response) > 0:
+
+            good_case = False
+
+            if "Accepted" in parsed_response or "Accepted" in parsed_response['SubmissionResult']:
+                self.logger.debug("Accepted response for url'" + url + "' Response:" + str(parsed_response))
+
+                if 'SubmissionResult' in parsed_response and 'Accepted' in parsed_response['SubmissionResult']:
+                    self.logger.info(
+                        "BlueCoat Accepted response for url'" + url + "' SubmissionResult (uniq id)=" + str(
```
Comment (cgi1) [inline intelmq/lib/BlueCoatAPI.py]: @wagner-certat: This logging is only possible with the baseLib. (And the rest of it in this libary)

**Comment by cgi1** [comment]
@wagner-certat: Nope i am talking about the logging in [intelmq/lib/BlueCoatAPI.py](https://github.com/certtools/intelmq/pull/969/files#diff-c844996a3901bd6b0bc83d48e0462ee3)

File: intelmq/lib/BlueCoatAPI.py (hunk 5 above)
Comment (wagner-certat) [inline intelmq/lib/BlueCoatAPI.py]: From the bot you can pass `self.logger` to the lib on initialization.

**Comment by wagner-certat** [comment]
Have a look at the xmpp bots, both are passing the logger instance to other classes: https://github.com/certtools/intelmq/blob/master/intelmq/bots/collectors/xmpp/collector.py

**Comment by wagner-certat** [comment]
Closing here now

**Pull request closed without merging**